File: erts_diff_drive.py
#!/usr/bin/env python3

"""
Runs Extended Rauch-Tung-Striebel controller on differential drive.
https://file.tavsys.net/control/papers/Extended%20Rauch-Tung-Striebel%20Controller%2C%20ZAGSS.pdf
"""

import math
import time
import sys

import frccontrol as fct
import matplotlib.pyplot as plt
import numpy as np

DT = 0.02
HORIZON = 100


def lerp(a, b, t):
    """
    Linear interpolation between a and b.

    Keyword arguments:
    a -- left value
    b -- right value
    t -- interpolant [0, 1]
    """
    return a + t * (b - a)


def get_square_refs():
    """
    Generate square path reference.
    """
    refs = []
    heading = 0.0
    points = [
        (0, 0),
        (2, 0),
        (2, 2),
        (-2, 2),
        (-2, 4),
        (4, 4),
        (4, -2),
        (0, -2),
    ]
    v = 2.0
    point_vectors = [np.array([[x], [y]]) for x, y in points]
    for pt0, pt1 in zip(point_vectors, point_vectors[1:]):
        diff = pt1 - pt0
        t = math.sqrt((diff.T @ diff)[0, 0]) / v
        new_heading = math.atan2(diff[1, 0], diff[0, 0])

        heading_diff = new_heading - heading
        if heading_diff > math.pi:
            if heading_diff > 0:
                heading += heading_diff - 2 * math.pi
            else:
                heading += heading_diff + 2 * math.pi
        else:
            heading += heading_diff

        num_pts = int(t / DT)
        for j in range(num_pts):
            mid = lerp(pt0, pt1, j / num_pts)
            ref = np.array(
                [
                    [mid[0, 0]],
                    [mid[1, 0]],
                    [heading],
                    [v],
                    [v],
                ]
            )
            refs.append(ref)
    return refs


class DifferentialDrive:
    """
    Differential drive with ERTS controller.
    """

    def __init__(self, dt):
        """Drivetrain subsystem.

        Keyword arguments:
        dt -- time between model/controller updates
        """
        self.dt = dt

        # Number of motors per side
        num_motors = 3
        # Gear ratio
        G = 60.0 / 11.0
        # Drivetrain mass in kg
        m = 52.0
        # Radius of wheels in meters
        r = 0.08255 / 2.0
        # Radius of robot in meters
        self.rb = 0.59055 / 2.0
        # Moment of inertia of the differential drive in kg-m²
        J = 6.0

        motor = fct.models.gearbox(fct.models.MOTOR_CIM, num_motors)

        C1 = -(G**2) * motor.Kt / (motor.Kv * motor.R * r**2)
        C2 = G * motor.Kt / (motor.R * r)
        C3 = -(G**2) * motor.Kt / (motor.Kv * motor.R * r**2)
        C4 = G * motor.Kt / (motor.R * r)
        self.velocity_A = np.array(
            [
                [(1 / m + self.rb**2 / J) * C1, (1 / m - self.rb**2 / J) * C3],
                [(1 / m - self.rb**2 / J) * C1, (1 / m + self.rb**2 / J) * C3],
            ]
        )
        self.velocity_B = np.array(
            [
                [(1 / m + self.rb**2 / J) * C2, (1 / m - self.rb**2 / J) * C4],
                [(1 / m - self.rb**2 / J) * C2, (1 / m + self.rb**2 / J) * C4],
            ]
        )

        # Linearized dynamics
        self.Ac = np.zeros((5, 5))
        self.Ac[2, 3] = -0.5 / self.rb
        self.Ac[2, 4] = 0.5 / self.rb
        self.Ac[3:5, 3:5] = self.velocity_A
        self.Bc = np.block(
            [
                [np.zeros((3, 2))],
                [self.velocity_B],
            ]
        )

        # Sim variables
        self.x = np.zeros((5, 1))
        self.u = np.zeros((2, 1))
        self.y = np.zeros((3, 1))

        self.u_min = np.array([[-12.0], [-12.0]])
        self.u_max = np.array([[12.0], [12.0]])

        # States: x (m), y (m), heading (rad), left velocity (m/s),
        #         right velocity (m/s)
        # Q = diag(1/q²)
        # Q⁻¹ = diag(q²)
        self.Qinv = np.diag(np.square([0.125, 0.125, 10.0, 0.95, 0.95]))

        # Inputs: Left voltage (V), right voltage (V)
        # R = diag(1/r²)
        # R⁻¹ = diag(r²)
        self.Rinv = np.diag(np.square([12.0, 12.0]))

        self.t = 0

        # Get reference trajectory
        self.refs = get_square_refs()

        # Kalman smoother storage
        self.x_hat_pre = [np.zeros((5, 1)) for _ in range(len(self.refs))]
        self.x_hat_post = [np.zeros((5, 1)) for _ in range(len(self.refs))]
        self.A = [np.zeros((5, 5)) for _ in range(len(self.refs))]
        self.P_pre = [np.zeros((5, 5)) for _ in range(len(self.refs))]
        self.P_post = [np.zeros((5, 5)) for _ in range(len(self.refs))]
        self.x_hat_smooth = [np.zeros((5, 1)) for _ in range(len(self.refs))]

    def f(self, x, u):
        """
        Nonlinear differential drive dynamics.

        States: [[x], [y], [heading], [left velocity], [right velocity]]
        Inputs: [[left voltage], [right voltage]]

        Keyword arguments:
        x -- state vector
        u -- input vector

        Returns:
        dx/dt -- state derivative
        """
        return np.block(
            [
                [(x[3, 0] + x[4, 0]) / 2.0 * math.cos(x[2, 0])],
                [(x[3, 0] + x[4, 0]) / 2.0 * math.sin(x[2, 0])],
                [(x[4, 0] - x[3, 0]) / (2.0 * self.rb)],
                [self.velocity_A @ x[3:5, :] + self.velocity_B @ u],
            ]
        )

    def h(self, x):
        """
        Nonlinear differential drive measurement model.

        Outputs: [[x], [y], [heading]]

        Keyword arguments:
        x -- state vector

        Returns:
        y -- measurement vector
        """
        return x[:3, :]

    # pragma pylint: disable=unused-argument
    def df_dx(self, x, u):
        """
        Returns the Jacobian of f with respect to the state.

        Keyword arguments:
        x -- the current state
        u -- the current input
        """
        v = (x[3, 0] + x[4, 0]) / 2.0
        c = math.cos(x[2, 0])
        s = math.sin(x[2, 0])
        self.Ac[0, 2] = -v * s
        self.Ac[0, 3] = 0.5 * c
        self.Ac[0, 4] = 0.5 * c
        self.Ac[1, 2] = v * c
        self.Ac[1, 3] = 0.5 * s
        self.Ac[1, 4] = 0.5 * s
        return self.Ac

    # pragma pylint: disable=unused-argument
    def dh_dx(self, x, u):
        """
        Returns the Jacobian of h with respect to the state.

        Keyword arguments:
        x -- the current state
        u -- the current input
        """
        return np.array([[1, 0, 0, 0, 0], [0, 1, 0, 0, 0], [0, 0, 1, 0, 0]])

    # pragma pylint: disable=unused-argument
    def update(self, r, next_r):
        """
        Advance the model by one timestep.

        Keyword arguments:
        r -- the current reference
        next_r -- the next reference
        """
        self.x = fct.rk4(self.f, self.x, self.u, self.dt)

        start = time.time()
        # Since this is the last reference, there are no reference dynamics to
        # follow
        if self.t == len(self.refs) - 1:
            self.u = np.zeros((2, 1))
            return

        states = self.x.shape[0]

        # Linearize model
        Ac = self.df_dx(self.x, np.zeros((2, 1)))
        _, B = fct.discretize_ab(Ac, self.Bc, self.dt)
        Binv = np.linalg.pinv(B)

        self.x_hat_pre[self.t] = self.x.copy()
        self.P_pre[self.t] = np.zeros((states, states))
        self.x_hat_post[self.t] = self.x.copy()
        self.P_post[self.t] = np.zeros((states, states))

        # Prediction
        N = min(len(self.refs) - 1, self.t + HORIZON)
        for τ in range(self.t + 1, N + 1):
            self.x_hat_pre[τ] = fct.rk4(
                self.f, self.x_hat_post[τ - 1], np.zeros((2, 1)), self.dt
            )

            # Linearization
            self.A[τ - 1] = fct.discretize_a(
                self.df_dx(self.x_hat_post[τ - 1], np.zeros((2, 1))), self.dt
            )
            C = self.dh_dx(self.x_hat_pre[τ], np.zeros((2, 1)))

            s_τ = C @ self.refs[τ]

            self.P_pre[τ] = (
                self.A[τ - 1] @ self.P_post[τ - 1] @ self.A[τ - 1].T
                + B @ self.Rinv @ B.T
            )

            # S = CPCᵀ + CQ⁻¹Cᵀ
            S = C @ self.P_pre[τ] @ C.T + C @ self.Qinv @ C.T

            # We want to put K = PCᵀS⁻¹ into Ax = b form so we can solve it more
            # efficiently.
            #
            # K = PCᵀS⁻¹
            # KS = PCᵀ
            # (KS)ᵀ = (PCᵀ)ᵀ
            # SᵀKᵀ = CPᵀ
            #
            # The solution of Ax = b can be found via x = A.solve(b).
            #
            # Kᵀ = Sᵀ.solve(CPᵀ)
            # K = Sᵀ.solve(CPᵀ)ᵀ
            K = np.linalg.solve(S.T, C @ self.P_pre[τ].T).T

            self.x_hat_post[τ] = self.x_hat_pre[τ] + K @ (
                s_τ - self.h(self.x_hat_pre[τ])
            )
            self.P_post[τ] = (np.eye(5) - K @ C) @ self.P_pre[τ] @ (
                np.eye(5) - K @ C
            ).T + K @ C @ self.Qinv @ C.T @ K.T

        # Last filtered estimate is already optimal smoothed estimate
        self.x_hat_smooth[N] = self.x_hat_post[N]

        # Smoothing
        for τ in range(N - 1, (self.t + 1) - 1, -1):
            # L = P⁺[τ] A[τ]ᵀ P⁻[τ + 1]⁻¹
            # L P⁻[τ + 1] = P⁺[τ] A[τ]ᵀ
            # P⁻[τ + 1]ᵀ Lᵀ = A[τ] P⁺[τ]ᵀ
            # Lᵀ = P⁻[τ + 1]ᵀ.solve(A[τ] P⁺[τ]ᵀ)
            # L = P⁻[τ + 1]ᵀ.solve(A[τ] P⁺[τ]ᵀ)ᵀ
            try:
                L = np.linalg.solve(self.P_pre[τ + 1].T, self.A[τ] @ self.P_post[τ].T).T
            except np.linalg.LinAlgError:
                L = self.P_post[τ] @ self.A[τ].T @ np.linalg.pinv(self.P_pre[τ + 1])
            self.x_hat_smooth[τ] = self.x_hat_post[τ] + L @ (
                self.x_hat_smooth[τ + 1] - self.x_hat_pre[τ + 1]
            )

        # x̂ₖ₊₁ = f(x̂ₖ) + Buₖ
        # Buₖ = x̂ₖ₊₁ − f(x̂ₖ)
        # uₖ = B⁺(x̂ₖ₊₁ − f(x̂ₖ))
        self.u = Binv @ (
            self.x_hat_smooth[self.t + 1]
            - fct.rk4(self.f, self.x_hat_post[self.t], np.zeros((2, 1)), self.dt)
        )

        u_cap = np.max(np.abs(self.u))
        if u_cap > 12.0:
            self.u = self.u / u_cap * 12.0
        self.t += 1

        end = time.time()
        print(
            f"\riteration: {self.t}/{len(self.refs) - 1}, dt = {round((end - start) * 1e3)} ms ",
            end="",
        )


def main():
    """Entry point."""
    refs = get_square_refs()

    t = [0.0]
    for _ in range(len(refs) - 1):
        t.append(t[-1] + DT)

    x = np.zeros((5, 1))
    diff_drive = DifferentialDrive(DT)
    diff_drive.x = x.copy()

    start = time.time()
    ref_rec, state_rec, u_rec, _ = fct.generate_time_responses(diff_drive, refs)
    end = time.time()
    print(f"\nTotal time = {round(end - start, 3)} s")

    plt.figure(1)
    plt.plot(ref_rec[0, :], ref_rec[1, :], label="Reference trajectory")
    plt.plot(state_rec[0, :], state_rec[1, :], label="ERTS controller")
    plt.xlabel("x (m)")
    plt.ylabel("y (m)")
    plt.legend()

    plt.gca().set_aspect(1.0)
    plt.gca().set_box_aspect(1.0)

    if "--noninteractive" in sys.argv:
        plt.savefig("erts_diff_drive_xy.png")

    fct.plot_time_responses(
        [
            "x position (m)",
            "y position (m)",
            "Heading (rad)",
            "Left velocity (m/s)",
            "Right velocity (m/s)",
        ],
        ["Left voltage (V)", "Right voltage (V)"],
        t,
        ref_rec,
        state_rec,
        u_rec,
    )

    if "--noninteractive" in sys.argv:
        plt.savefig("erts_diff_drive_response.png")
    else:
        plt.show()


if __name__ == "__main__":
    main()

File: src/ERTSDiffDrive.cpp
// Copyright (c) Tyler Veness

// Runs Extended Rauch-Tung-Striebel controller on differential drive.
// https://file.tavsys.net/control/papers/Extended%20Rauch-Tung-Striebel%20Controller%2C%20ZAGSS.pdf

#include <algorithm>
#include <array>
#include <chrono>
#include <cmath>
#include <numbers>
#include <print>
#include <string_view>
#include <vector>

#include <Eigen/Cholesky>
#include <Eigen/Core>
#include <Eigen/QR>

#include "DCMotor.hpp"
#include "Discretization.hpp"
#include "RK4.hpp"
#include "WriteVectorCSV.hpp"

using namespace std::chrono_literals;
using namespace std::string_view_literals;

constexpr std::chrono::duration<double> DT = 20ms;
constexpr int HORIZON = 100;

/**
 * Linear interpolation between a and b.
 *
 * @param a Left value.
 * @param b Right value.
 * @param t Interpolant [0, 1].
 */
template <typename T>
T lerp(const T& a, const T& b, double t) {
  return a + t * (b - a);
}

/**
 * Generate square path reference.
 */
std::vector<Eigen::Vector<double, 5>> GetSquareRefs() {
  std::vector<Eigen::Vector<double, 5>> refs;

  double heading = 0.0;
  constexpr double v = 2.0;
  std::vector<Eigen::Vector<double, 2>> point_vectors{
      {0, 0}, {2, 0}, {2, 2}, {-2, 2}, {-2, 4}, {4, 4}, {4, -2}, {0, -2}};

  for (size_t i = 0; i < point_vectors.size() - 1; ++i) {
    const auto& pt0 = point_vectors[i];
    const auto& pt1 = point_vectors[i + 1];

    Eigen::Vector<double, 2> diff = pt1 - pt0;
    double t = std::sqrt((diff.transpose() * diff)(0)) / v;
    double new_heading = std::atan2(diff(1), diff(0));

    double heading_diff = new_heading - heading;
    if (heading_diff > std::numbers::pi) {
      if (heading_diff > 0) {
        heading += heading_diff - 2 * std::numbers::pi;
      } else {
        heading += heading_diff + 2 * std::numbers::pi;
      }
    } else {
      heading += heading_diff;
    }

    int num_pts = t / DT.count();
    for (int j = 0; j < num_pts; ++j) {
      Eigen::Vector<double, 2> mid =
          lerp(pt0, pt1, static_cast<double>(j) / num_pts);
      refs.emplace_back(
          Eigen::Vector<double, 5>{{mid(0), mid(1), heading, v, v}});
    }
  }

  return refs;
}

/**
 * Differential drive with ERTS controller.
 */
class DifferentialDrive {
 public:
  // Number of motors per side
  static constexpr int num_motors = 3;

  // Gear ratio
  static constexpr double G = 60.0 / 11.0;

  // Drivetrain mass in kg
  static constexpr double m = 52.0;

  // Radius of wheels in meters
  static constexpr double r = 0.08255 / 2.0;

  // Radius of robot in meters
  static constexpr double rb = 0.59055 / 2.0;

  // Moment of inertia of the differential drive in kg-m²
  static constexpr double J = 6.0;

  static constexpr DCMotor motor = DCMotor::CIM(num_motors);

  static constexpr double C1 =
      -(G * G) * motor.Kt / (motor.Kv * motor.R * r * r);
  static constexpr double C2 = G * motor.Kt / (motor.R * r);
  static constexpr double C3 =
      -(G * G) * motor.Kt / (motor.Kv * motor.R * r * r);
  static constexpr double C4 = G * motor.Kt / (motor.R * r);

  static constexpr Eigen::Matrix<double, 2, 2> velocity_A{{
      {(1 / m + rb * rb / J) * C1, (1 / m - rb * rb / J) * C3},
      {(1 / m - rb * rb / J) * C1, (1 / m + rb * rb / J) * C3},
  }};
  static constexpr Eigen::Matrix<double, 2, 2> velocity_B{{
      {(1 / m + rb * rb / J) * C2, (1 / m - rb * rb / J) * C4},
      {(1 / m - rb * rb / J) * C2, (1 / m + rb * rb / J) * C4},
  }};

  // States: x (m), y (m), heading (rad), left velocity (m/s),
  //         right velocity (m/s)
  // Q = diag(1/q²)
  // Q⁻¹ = diag(q²)
  static constexpr Eigen::Matrix<double, 5, 5> Qinv{{0.125 * 0.125, 0, 0, 0, 0},
                                                    {0, 0.125 * 0.125, 0, 0, 0},
                                                    {0, 0, 10.0 * 10.0, 0, 0},
                                                    {0, 0, 0, 0.95 * 0.95, 0},
                                                    {0, 0, 0, 0, 0.95 * 0.95}};

  // Inputs: Left voltage (V), right voltage (V)
  // R = diag(1/r²)
  // R⁻¹ = diag(r²)
  static constexpr Eigen::Matrix<double, 2, 2> Rinv{{12.0 * 12.0, 0},
                                                    {0, 12.0 * 12.0}};

  std::chrono::duration<double> m_dt = 0s;

  size_t t = 0;

  // Sim variables
  Eigen::Vector<double, 5> x = Eigen::Vector<double, 5>::Zero();
  Eigen::Vector<double, 2> u = Eigen::Vector<double, 2>::Zero();

  std::vector<Eigen::Vector<double, 5>> m_refs;

  // Kalman smoother storage
  std::vector<Eigen::Vector<double, 5>> x_hat_pre;
  std::vector<Eigen::Vector<double, 5>> x_hat_post;
  std::vector<Eigen::Matrix<double, 5, 5>> A;
  std::vector<Eigen::Matrix<double, 5, 5>> P_pre;
  std::vector<Eigen::Matrix<double, 5, 5>> P_post;
  std::vector<Eigen::Vector<double, 5>> x_hat_smooth;

  static constexpr Eigen::Vector<double, 2> u_min{{-12.0}, {-12.0}};
  static constexpr Eigen::Vector<double, 2> u_max{{12.0}, {12.0}};

  /**
   * Drivetrain subsystem.
   *
   * @param dt Time between model/controller updates.
   */
  explicit DifferentialDrive(std::chrono::duration<double> dt) : m_dt{dt} {
    // Get reference trajectory
    m_refs = GetSquareRefs();

    // Kalman smoother storage
    x_hat_pre.reserve(m_refs.size());
    x_hat_post.reserve(m_refs.size());
    A.reserve(m_refs.size());
    P_pre.reserve(m_refs.size());
    P_post.reserve(m_refs.size());
    x_hat_smooth.reserve(m_refs.size());
    for (size_t i = 0; i < m_refs.size(); ++i) {
      x_hat_pre.emplace_back(Eigen::Vector<double, 5>::Zero());
      x_hat_post.emplace_back(Eigen::Vector<double, 5>::Zero());
      A.emplace_back(Eigen::Matrix<double, 5, 5>::Zero());
      P_pre.emplace_back(Eigen::Matrix<double, 5, 5>::Zero());
      P_post.emplace_back(Eigen::Matrix<double, 5, 5>::Zero());
      x_hat_smooth.emplace_back(Eigen::Vector<double, 5>::Zero());
    }
  }

  /**
   * Nonlinear differential drive dynamics.
   *
   * States: [[x], [y], [heading], [left velocity], [right velocity]]
   * Inputs: [[left voltage], [right voltage]]
   *
   * @param x State vector.
   * @param u Input vector.
   */
  Eigen::Vector<double, 5> f(const Eigen::Vector<double, 5>& x,
                             const Eigen::Vector<double, 2>& u) {
    Eigen::Vector<double, 5> dxdt;

    double v = (x(3) + x(4)) / 2.0;
    dxdt(0) = v * std::cos(x(2));
    dxdt(1) = v * std::sin(x(2));
    dxdt(2) = (x(4) - x(3)) / (2.0 * rb);
    dxdt.segment<2>(3) = velocity_A * x.segment<2>(3) + velocity_B * u;

    return dxdt;
  }

  /**
   * Nonlinear differential drive measurement model.
   *
   * Outputs: [[x], [y], [heading]]
   *
   * @param x State vector.
   */
  Eigen::Vector<double, 3> h(const Eigen::Vector<double, 5>& x) const {
    return x.segment<3>(0);
  }

  /**
   * Returns the Jacobian of f with respect to the state.
   *
   * @param x The current state.
   * @param u The current input.
   */
  Eigen::Matrix<double, 5, 5> df_dx(
      const Eigen::Vector<double, 5>& x,
      [[maybe_unused]] const Eigen::Vector<double, 2>& u) {
    double v = (x(3) + x(4)) / 2.0;
    double c = std::cos(x(2));
    double s = std::sin(x(2));

    return Eigen::Matrix<double, 5, 5>{
        {0.0, 0.0, -v * s, 0.5 * c, 0.5 * c},
        {0.0, 0.0, v * c, 0.5 * s, 0.5 * s},
        {0.0, 0.0, 0.0, -0.5 / rb, 0.5 / rb},
        {0.0, 0.0, 0.0, velocity_A(0, 0), velocity_A(0, 1)},
        {0.0, 0.0, 0.0, velocity_A(1, 0), velocity_A(1, 1)}};
  }

  /**
   * Returns the Jacobian of f with respect to the input.
   *
   * @param x The current state.
   * @param u The current input.
   */
  Eigen::Matrix<double, 5, 2> df_du(
      [[maybe_unused]] const Eigen::Vector<double, 5>& x,
      [[maybe_unused]] const Eigen::Vector<double, 2>& u) {
    return Eigen::Matrix<double, 5, 2>{
        {0.0, 0.0},
        {0.0, 0.0},
        {0.0, 0.0},
        {(1 / m + rb * rb / J) * C2, (1 / m - rb * rb / J) * C4},
        {(1 / m - rb * rb / J) * C2, (1 / m + rb * rb / J) * C4}};
  }

  /**
   * Returns the Jacobian of h with respect to the state.
   *
   * @param x The current state.
   * @param u The current input.
   */
  Eigen::Matrix<double, 3, 5> dh_dx(
      [[maybe_unused]] const Eigen::Vector<double, 5>& x,
      [[maybe_unused]] const Eigen::Vector<double, 2>& u) {
    return Eigen::Matrix<double, 3, 5>{
        {1, 0, 0, 0, 0}, {0, 1, 0, 0, 0}, {0, 0, 1, 0, 0}};
  }

  /**
   * Advance the model by one timestep.
   *
   * @param r The current reference.
   * @param nextR The next reference.
   */
  void Update([[maybe_unused]] const Eigen::Vector<double, 5>& r,
              [[maybe_unused]] const Eigen::Vector<double, 5>& nextR) {
    x = RK4([this](const auto& x, const auto& u) { return f(x, u); }, x, u,
            m_dt);

    // Since this is the last reference, there are no
    // reference dynamics to follow
    if (t == m_refs.size() - 1) {
      u = Eigen::Vector<double, 2>::Zero();
      return;
    }

    // Linearize model
    Eigen::Matrix<double, 5, 5> Ac = df_dx(x, Eigen::Vector<double, 2>::Zero());
    Eigen::Matrix<double, 5, 2> Bc = df_du(x, Eigen::Vector<double, 2>::Zero());
    Eigen::Matrix<double, 5, 5> Ad;
    Eigen::Matrix<double, 5, 2> Bd;
    DiscretizeAB<5, 2>(Ac, Bc, m_dt, &Ad, &Bd);

    x_hat_pre[t] = x;
    P_pre[t] = Eigen::Matrix<double, 5, 5>::Zero();
    x_hat_post[t] = x;
    P_post[t] = Eigen::Matrix<double, 5, 5>::Zero();

    // Prediction
    int N = std::min<double>(m_refs.size() - 1, t + HORIZON);
    for (int τ = t + 1; τ < N + 1; ++τ) {
      x_hat_pre[τ] =
          RK4([this](const auto& x, const auto& u) { return f(x, u); },
              x_hat_post[τ - 1], Eigen::Vector<double, 2>::Zero(), m_dt);

      // Linearization
      DiscretizeA<5>(df_dx(x_hat_post[τ - 1], Eigen::Vector<double, 2>::Zero()),
                     m_dt, &A[τ - 1]);

      Eigen::Matrix<double, 3, 5> C =
          dh_dx(x_hat_post[τ - 1], Eigen::Vector<double, 2>::Zero());
      Eigen::Matrix<double, 3, 1> s_τ = C * m_refs[τ];

      P_pre[τ] = A[τ - 1] * P_post[τ - 1] * A[τ - 1].transpose() +
                 Bd * Rinv * Bd.transpose();

      // S = CPCᵀ + CQ⁻¹Cᵀ
      Eigen::Matrix<double, 3, 3> S =
          C * P_pre[τ] * C.transpose() + C * Qinv * C.transpose();

      // We want to put K = PCᵀS⁻¹ into Ax = b form so we can solve it more
      // efficiently.
      //
      // K = PCᵀS⁻¹
      // KS = PCᵀ
      // (KS)ᵀ = (PCᵀ)ᵀ
      // SᵀKᵀ = CPᵀ
      //
      // The solution of Ax = b can be found via x = A.solve(b).
      //
      // Kᵀ = Sᵀ.solve(CPᵀ)
      // K = (Sᵀ.solve(CPᵀ))ᵀ
      Eigen::Matrix<double, 5, 3> K =
          S.transpose().llt().solve(C * P_pre[τ].transpose()).transpose();

      x_hat_post[τ] = x_hat_pre[τ] + K * (s_τ - h(x_hat_pre[τ]));

      constexpr Eigen::Matrix<double, 5, 5> I{{1, 0, 0, 0, 0},
                                              {0, 1, 0, 0, 0},
                                              {0, 0, 1, 0, 0},
                                              {0, 0, 0, 1, 0},
                                              {0, 0, 0, 0, 1}};
      P_post[τ] = (I - K * C) * P_pre[τ] * (I - K * C).transpose() +
                  K * C * Qinv * C.transpose() * K.transpose();
    }

    // Last filtered estimate is already optimal smoothed estimate
    x_hat_smooth[N] = x_hat_post[N];

    // Smoothing
    for (size_t τ = N - 1; τ > (t + 1) - 1; --τ) {
      // L = P⁺[τ] A[τ]ᵀ P⁻[τ + 1]⁻¹
      // L P⁻[τ + 1] = P⁺[τ] A[τ]ᵀ
      // P⁻[τ + 1]ᵀ Lᵀ = A[τ] P⁺[τ]ᵀ
      // Lᵀ = P⁻[τ + 1]ᵀ.solve(A[τ] P⁺[τ]ᵀ)
      // L = P⁻[τ + 1]ᵀ.solve(A[τ] P⁺[τ]ᵀ)ᵀ
      Eigen::Matrix<double, 5, 5> L = P_pre[τ + 1]
                                          .transpose()
                                          .llt()
                                          .solve(A[τ] * P_post[τ].transpose())
                                          .transpose();

      x_hat_smooth[τ] =
          x_hat_post[τ] + L * (x_hat_smooth[τ + 1] - x_hat_pre[τ + 1]);
    }

    // x̂ₖ₊₁ = f(x̂ₖ) + Buₖ
    // Buₖ = x̂ₖ₊₁ − f(x̂ₖ)
    // uₖ = B⁺(x̂ₖ₊₁ − f(x̂ₖ))
    u = Bd.householderQr().solve(
        x_hat_smooth[t + 1] -
        RK4([this](const auto& x, const auto& u) { return f(x, u); },
            x_hat_post[t], Eigen::Vector<double, 2>::Zero(), m_dt));

    double u_cap = u.lpNorm<Eigen::Infinity>();
    if (u_cap > 12.0) {
      u = u / u_cap * 12.0;
    }
    ++t;
  }
};

int main() {
  auto refs = GetSquareRefs();

  std::vector<double> t;
  t.emplace_back(0.0);
  for (size_t i = 0; i < refs.size() - 1; ++i) {
    t.emplace_back(t.back() + DT.count());
  }

  Eigen::Vector<double, 5> x = Eigen::Vector<double, 5>::Zero();
  auto diff_drive = DifferentialDrive(DT);
  diff_drive.x = x;

  auto start = std::chrono::system_clock::now();

  std::vector<Eigen::Vector<double, 5>> r_rec;
  r_rec.reserve(refs.size());
  std::vector<Eigen::Vector<double, 5>> x_rec;
  x_rec.reserve(refs.size());
  std::vector<Eigen::Vector<double, 2>> u_rec;
  u_rec.reserve(refs.size());

  // Run simulation
  for (size_t i = 0; i < refs.size(); ++i) {
    if (i < refs.size() - 1) {
      diff_drive.Update(refs[i], refs[i + 1]);
    } else {
      diff_drive.Update(refs[i], refs[i]);
    }

    // Log states for plotting
    r_rec.emplace_back(refs[i]);
    x_rec.emplace_back(diff_drive.x);
    u_rec.emplace_back(diff_drive.u);
  }

  auto end = std::chrono::system_clock::now();

  std::print(stderr, "Total time = {} ms\n",
             std::chrono::duration_cast<std::chrono::microseconds>(end - start)
                     .count() /
                 1e3);

  WriteVectorCSV("ERTS references.csv"sv, t, r_rec,
                 std::array{"Time (s)"sv, "X position (m)"sv,
                            "Y position (m)"sv, "Heading (rad)"sv,
                            "Left velocity (m/s)"sv, "Right velocity (m/s)"sv});

  WriteVectorCSV("ERTS states.csv"sv, t, x_rec,
                 std::array{"Time (s)"sv, "X position (m)"sv,
                            "Y position (m)"sv, "Heading (rad)"sv,
                            "Left velocity (m/s)"sv, "Right velocity (m/s)"sv});

  WriteVectorCSV(
      "ERTS inputs.csv"sv, t, u_rec,
      std::array{"Time (s)"sv, "Left voltage (V)"sv, "Right voltage (V)"sv});
}
